mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/query/nativequery/MongoDBNativeQueryPaginationTest.java

```diff
@@ -0,0 +1,253 @@
+import org.junit.Test;
+
+/**
+ * Test the pagination for native queries with MongoDB
+ *
+ * @author Fabio Massimo Ercoli
+ */
+public class MongoDBNativeQueryPaginationTest extends OgmTestCase {
+
+	public static final String QUERY_FIND = "db.WILDE_POEM.find( { 'author' : 'Oscar Wilde' } )";
+	public static final String QUERY_AGGREGATE_MATCH = "db.WILDE_POEM.aggregate( [ { '$match' : { 'author' : 'Oscar Wilde' } } ] )";
+
+	private final OscarWildePoem portia = new OscarWildePoem( 1L, "Portia", "Oscar Wilde", 1881 );
+	private final OscarWildePoem athanasia = new OscarWildePoem( 2L, "Athanasia", "Oscar Wilde", 1879 );
+	private final OscarWildePoem imperatrix = new OscarWildePoem( 3L, "Ave Imperatrix", "Oscar Wilde", 1882 );
+	private final OscarWildePoem intellectualis = new OscarWildePoem( 4L, "Amor Intellectualis", "Oscar Wilde", 1881 );
+	private final OscarWildePoem apologia = new OscarWildePoem( 5L, "Apologias ", "Oscar Wilde", 1881 );
+	private final OscarWildePoem easter = new OscarWildePoem( 6L, "Easter Day", "Oscar Wilde", 1881 );
+	private final OscarWildePoem rome = new OscarWildePoem( 7L, "Rome Unvisited", "Oscar Wilde", 1881 );
+	private final OscarWildePoem miniato = new OscarWildePoem( 8L, "San Miniato", "Oscar Wilde", 1881 );
+	private final OscarWildePoem liberty = new OscarWildePoem( 9L, "Sonnet to Liberty", "Oscar Wilde", 1881 );
+	private final OscarWildePoem vita = new OscarWildePoem( 10L, "Vita Nuova", "Oscar Wilde", 1881 );
+
+	private final OscarWildePoem[] poems = { portia, athanasia, imperatrix, intellectualis, apologia, easter, rome, miniato, liberty, vita };
+
+	@Before
+	public void init() {
+		inTransaction( session -> {
+			for ( OscarWildePoem poem : poems ) {
+				session.persist( poem );
+			}
+		} );
+	}
+
+	@After
+	public void tearDown() {
+		inTransaction( session -> {
+			for ( OscarWildePoem poem : poems ) {
+				delete( session, poem );
+			}
+		} );
+	}
+
+	private void delete(Session session, OscarWildePoem poem) {
+		Object entity = session.get( OscarWildePoem.class, poem.getId() );
+		if ( entity != null ) {
+			session.delete( entity );
+		}
+	}
+
+	@Test
+	public void testFindPage1Size10() {
+		inTransaction( session -> {
+			List result = session.createNativeQuery( QUERY_FIND )
+				.addEntity( OscarWildePoem.TABLE_NAME, OscarWildePoem.class )
+				.setFirstResult( 0 )
+				.setMaxResults( 10 )
+				.list();
+
+			assertThat( result.size() ).isEqualTo( 10 );
+
+			new PoemResultValidator( poems )
+				.validateAndRemove( result )
+				.mustBeEmpty();
+		} );
+	}
+
+	@Test
+	public void testAggregationPipelinePage1Size10() {
+		inTransaction( session -> {
+			List result = session.createNativeQuery( QUERY_AGGREGATE_MATCH )
+				.addEntity( OscarWildePoem.TABLE_NAME, OscarWildePoem.class )
+				.setFirstResult( 0 )
+				.setMaxResults( 10 )
+				.list();
+
+			assertThat( result.size() ).isEqualTo( 10 );
+
+			new PoemResultValidator( poems )
+				.validateAndRemove( result )
```

Comment:
I don't think I like this approach. The problem is that I don't understand what's going on and which value you expect.
Also, I don't know what kind of validation you are applying. Don't you think this is simpler and easier to understand?

```
assertThat( result ).containsOnly(portia, athanasa, imperatrix);
```
This will also check the size.

I woudl also return less results so that it's easier to figure out what's missing

Reply:
There is another problem with this test, if the query returns the results from the wrong page I don't think it will fail. For example, it returns the results from the second page instead of the first one

Reply:
There is no order defined on the query, any result sequence should be be ok.

Reply:
I'm +1 with Davide. Add an order and test the results with `containsOnly()`. It's the only way to be sure everything works OK.

Reply:
You're right. I'm going to push the fix.